test/e2e/reconciler/apiexportendpointslice/apiexportendpointslice_test.go

```diff
@@ -0,0 +1,269 @@
+		return len(s.Status.APIExportEndpoints) == 1
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expecting a single endpoint")
+
+	t.Logf("Updating the previously created shard")
+	shard.Labels["region"] = "doesnotexist"
+	shard, err = shardClient.Cluster(core.RootCluster.Path()).Update(ctx, shard, metav1.UpdateOptions{})
+	require.NoError(t, err, "error updating Shard")
+
+	require.Eventually(t, func() bool {
+		s, err := e.kcpClusterClient.Cluster(e.partitionClusterPath).ApisV1alpha1().APIExportEndpointSlices().Get(ctx, slice.Name, metav1.GetOptions{})
+		require.NoError(t, err)
+		return len(s.Status.APIExportEndpoints) == 0
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expecting no endpoint")
+
+	t.Logf("Setting back the correct label")
+	shard.Labels["region"] = "apiexportendpointslice-test-region"
+	shard, err = shardClient.Cluster(core.RootCluster.Path()).Update(ctx, shard, metav1.UpdateOptions{})
+	require.NoError(t, err, "error updating Shard")
+
+	require.Eventually(t, func() bool {
+		s, err := e.kcpClusterClient.Cluster(e.partitionClusterPath).ApisV1alpha1().APIExportEndpointSlices().Get(ctx, slice.Name, metav1.GetOptions{})
+		require.NoError(t, err)
+		return len(s.Status.APIExportEndpoints) == 1
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expecting a single endpoint")
+
+	t.Logf("Deleting the shard")
+	err = shardClient.Cluster(core.RootCluster.Path()).Delete(ctx, shard.Name, metav1.DeleteOptions{})
+	require.NoError(t, err, "error deleting Shard")
+
+	require.Eventually(t, func() bool {
+		s, err := e.kcpClusterClient.Cluster(e.partitionClusterPath).ApisV1alpha1().APIExportEndpointSlices().Get(ctx, slice.Name, metav1.GetOptions{})
+		require.NoError(t, err)
+		return len(s.Status.APIExportEndpoints) == 0
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expecting no endpoint")
+
+	t.Logf("Creating a slice without partition")
+	sliceWithAll := &apisv1alpha1.APIExportEndpointSlice{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "my-slice-without-partition",
+		},
+		Spec: apisv1alpha1.APIExportEndpointSliceSpec{
+			APIExport: apisv1alpha1.ExportBindingReference{
+				Path: e.exportClusterPath.String(),
+				Name: slice.Spec.APIExport.Name,
+			},
+		},
+	}
+	sliceClient := e.kcpClusterClient.ApisV1alpha1().APIExportEndpointSlices()
+	sliceWithAll, err = sliceClient.Cluster(e.partitionClusterPath).Create(ctx, sliceWithAll, metav1.CreateOptions{})
+	require.NoError(t, err, "error creating APIExportEndpointSlice")
+
+	require.Eventually(t, func() bool {
+		sliceWithAll, err := e.kcpClusterClient.Cluster(e.partitionClusterPath).ApisV1alpha1().APIExportEndpointSlices().Get(ctx, sliceWithAll.Name, metav1.GetOptions{})
+		require.NoError(t, err)
+		return len(sliceWithAll.Status.APIExportEndpoints) == 1
+	}, wait.ForeverTestTimeout, 100*time.Millisecond, "expecting a single endpoint for the root shard, got %d", len(sliceWithAll.Status.APIExportEndpoints))
+}
+
+func TestAPIExportEndpointSlice(t *testing.T) {
```

Comment:
have a hard time what these tests actually do. How is TestAPIExportEndpointSlice and TestAPIExportEndpointSlicePrivate different?

Hiding the core test logic in helpers is an anti-pattern. Build helpers (DSL-like) to build up the test setup, but put the actual thing tested here as high level as possible.

Reply:
And this core test code should read like a story, i.e. not full of helpers again, rather this should be a series of steps plus `require.SomeThing` lines.

Reply:
TestAPIExportEndpointSlicePrivate manipulates Shards, hence the need for a private cluster. It requires however the APIExport, APIExportEndpointSlice and other resources. The building of these resources is what is tested in TestAPIExportEndpointSlice. Lukasz explicitly asked for the tests to be run on both: a private and a shared cluster (for the part compatible with it). Without this ask TestAPIExportEndpointSlice would not be needed.